examples/medplum-patient-intake-demo/src/bots/core/intake-utils.ts

```diff
@@ -42,56 +57,166 @@ export const observationCategoryMapping: Record<string, CodeableConcept> = {
+        system: HTTP_TERMINOLOGY_HL7_ORG + '/CodeSystem/v3-ActCode',
+        code: 'nopp',
+        display: 'Notice of Privacy Practices',
+      },
+    ],
+  },
+  pay: {
+    coding: [
+      {
+        system: HTTP_TERMINOLOGY_HL7_ORG + '/CodeSystem/v3-ActCode',
+        code: 'pay',
+        display: 'Payment',
+      },
+    ],
+  },
+  med: {
+    coding: [
+      {
+        system: HTTP_TERMINOLOGY_HL7_ORG + '/CodeSystem/v3-ActCode',
+        code: 'med',
+        display: 'Medical',
+      },
+    ],
+  },
+};
+
+export const consentPolicyRuleMapping: Record<string, CodeableConcept> = {
+  hipaaNpp: {
+    coding: [
+      {
+        system: HTTP_TERMINOLOGY_HL7_ORG + '/CodeSystem/consentpolicycodes',
+        code: 'hipaa-npp',
+        display: 'HIPAA Notice of Privacy Practices',
+      },
+    ],
+  },
+  hipaaSelfPay: {
+    coding: [
+      {
+        system: HTTP_TERMINOLOGY_HL7_ORG + '/CodeSystem/consentpolicycodes',
+        code: 'hipaa-self-pay',
+        display: 'HIPAA Self-Pay Restriction',
+      },
+    ],
+  },
+};
+
+type ObservationQuestionnaireItemType = 'valueCodeableConcept' | 'valueDateTime';
+
+/**
+ * This function takes data about an Observation and creates or updates an existing
+ * resource with the same patient and code.
+ *
+ * @param medplum - A Medplum client
+ * @param patient - A Patient resource that will be stored as the subject
+ * @param code - A code for the observation
+ * @param category - A category for the observation
+ * @param answerType - The value[x] field where the answer should be stored
+ * @param value - The value to be stored in the observation
+ */
 export async function upsertObservation(
   medplum: MedplumClient,
   patient: Patient,
   code: CodeableConcept,
   category: CodeableConcept,
-  valueCoding: QuestionnaireResponseItemAnswer | undefined
+  answerType: ObservationQuestionnaireItemType,
+  value: QuestionnaireResponseItemAnswer | undefined
 ): Promise<void> {
   const coding = code.coding?.[0];
 
-  if (!valueCoding || !coding) {
+  if (!value || !coding) {
     return;
   }
 
-  const observation = createObservation(patient, code, category, valueCoding);
-
-  await medplum.upsertResource(observation, {
-    code: `${coding.system}|${coding.code}`,
-    subject: getReferenceString(patient),
-  });
-}
-
-function createObservation(
-  patient: Patient,
-  coding: CodeableConcept,
-  category: CodeableConcept,
-  valueCoding: Coding
-): Observation {
-  return {
+  const observation: Observation = {
     resourceType: 'Observation',
     status: 'final',
     subject: createReference(patient),
-    code: coding,
+    code: code,
     category: [category],
-    valueCodeableConcept: {
-      coding: [valueCoding],
-    },
+    ...(answerType === 'valueCodeableConcept'
+      ? {
+          valueCodeableConcept: {
+            coding: [value],
+          },
+        }
+      : {}),
+    ...(answerType === 'valueDateTime' ? { valueDateTime: value.valueDateTime } : {}),
```

Comment:
For readability, you should use an if statement _after_ this initial construction, rather than using the big ternary operator